### PR TITLE
[Labs/TagInput] Add index-aware onKeyDown and onKeyUp

### LINKED ISSUE #1801: [TagInput] index-aware `onKeyDown` handler
Would like `<TagInput onKeyDown>` that works like `onRemove`: receives event and index of active tag (or `undefined` if in the input). Probably also provide `onKeyUp` for completeness.

This would enable lots of sweet custom `TagInput` use cases like query building: start typing to insert an editable tag between two existing ones.

cc @ethanbond

### PR DESCRIPTION
**Pull request by cmslewis**
#### Checklist

- [x] Include tests
- [x] Update documentation

#### Changes proposed in this pull request:

- `TagInput` now supports `onKeyDown` and `onKeyUp` that receive the active tag index—or `undefined` if the input is focused.

### CODE REVIEW

**Comment by cmslewis** [comment]
@giladgray @llorca Tests should be fixed as of the latest commit. Take a look?

File: packages/labs/src/components/tag-input/tagInput.tsx [hunk 1]
```diff
@@ -303,9 +333,23 @@ export class TagInput extends AbstractComponent<ITagInputProps, ITagInputState>
                 this.handleBackspaceToRemove(event);
             }
         }
+
+        // stop the event from propagating to the container's onKeyDown
+        // callback. we'll invoke the top-level callback directly with an
+        // `undefined` index.
+        event.stopPropagation();
```
Comment (giladgray) [inline packages/labs/src/components/tag-input/tagInput.tsx]: might be safer to `preventDefault()` and check `isDefaultPrevented()` in `handleContainerKeyDown`?

File: packages/labs/src/components/tag-input/tagInput.tsx [hunk 2]
```diff
@@ -303,9 +333,23 @@ export class TagInput extends AbstractComponent<ITagInputProps, ITagInputState>
...
+
+        Utils.safeInvoke(this.props.onKeyDown, event, undefined);
         Utils.safeInvoke(this.props.inputProps.onKeyDown, event);
     };
 
+    private handleInputKeyUp = (event: React.KeyboardEvent<HTMLInputElement>) => {
+        // again, stop propagation to the container's onKeyUp callback.
+        event.stopPropagation();
```
Comment (giladgray) [inline packages/labs/src/components/tag-input/tagInput.tsx]: and again with the `preventDefault` suggestion

File: packages/labs/src/components/tag-input/tagInput.tsx [hunk 3]
```diff
@@ -303,9 +339,23 @@ export class TagInput extends AbstractComponent<ITagInputProps, ITagInputState>
...
+        // callback. we'll invoke the top-level callback directly with an
+        // `undefined` index.
+        event.preventDefault();
+
+        Utils.safeInvoke(this.props.onKeyDown, event, undefined);
         Utils.safeInvoke(this.props.inputProps.onKeyDown, event);
     };
 
+    private handleInputKeyUp = (event: React.KeyboardEvent<HTMLInputElement>) => {
+        // again, stop propagation to the container's onKeyUp callback.
```
Comment (giladgray) [inline packages/labs/src/components/tag-input/tagInput.tsx]: language

File: packages/labs/src/components/tag-input/tagInput.tsx [hunk 4]
```diff
@@ -303,9 +339,23 @@ export class TagInput extends AbstractComponent<ITagInputProps, ITagInputState>
                 this.handleBackspaceToRemove(event);
             }
         }
+
+        // stop the event from propagating to the container's onKeyDown
```
Comment (giladgray) [inline packages/labs/src/components/tag-input/tagInput.tsx]: language

File: packages/labs/src/components/tag-input/tagInput.tsx [hunk 5]
```diff
@@ -264,6 +285,21 @@ export class TagInput extends AbstractComponent<ITagInputProps, ITagInputState>
...
+            return;
+        }
+        Utils.safeInvoke(this.props.onKeyDown, event, this.state.activeIndex);
+    };
+
+    private handleContainerKeyUp = (event: React.KeyboardEvent<HTMLDivElement>) => {
+        if (event.isDefaultPrevented()) {
+            return;
+        }
+        Utils.safeInvoke(this.props.onKeyUp, event, this.state.activeIndex);
```
Comment (giladgray) [inline packages/labs/src/components/tag-input/tagInput.tsx]: I wonder if this makes more sense as
```
if (!prevented) then safeInvoke
```

**Pull request merged**